README.md
# pybatchexecute

> Python package to ease interactions with Google's `batchexecute` batch RPC system. It assists in preparing requests for it and decoding responses returned from it.

This package codifies the research of Ryan Kovatch[^1] and Boudewijn van Groos[^2] about Google's `/batchexecute` endpoint used accross Google Web applications.

## Usage

### Prepare a request

```python
>>> from pybatchexecute import PreparedBatchExecute
>>>
>>> # RPCs to call
>>> rpc1 = {"rpcid": "rpc1id", "args": ["some", "args"]}
>>> rpc2 = {"rpcid": "rpc2id", "args": ["some", "args"]}
>>>
>>> # Prepare request
>>> pbe = PreparedBatchExecute(
...     rpcs=[rpc1, rpc2],
...     host="example.com",
...     app="example",
... )
>>>
>>> # Access request attributes
>>> pbe.url
'https://example.com/_/example/data/batchexecute'
>>> pbe.headers
{'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8'}
>>> pbe.params
{'rpcids': 'rpc1id,rpc2id', '_reqid': 3807}
>>> pbe.data
{'f.req': '[[["rpc1id","[\\"some\\",\\"args\\"]",null,"1"],["rpc2id","[\\"some\\",\\"args\\"]",null,"2"]]]'}
```

Use the `url`, `headers`, `params` and `data` attributes (optionally combined with your own, e.g. an `at` parameter[^1] or a `Cookie` header[^1] for authenticated requests) to make a POST request with the HTTP library of your choice (e.g. [`requests`](https://requests.readthedocs.io/en/latest/)).

## Decode a response

```python
>>> from pybatchexecute import decode
>>>
>>> # Raw response
>>> # (from a request made with no `rt` parameter)
>>> raw_reponse = r"""
... )]}'
...
... [["wrb.fr","rpc1id","[\"some\",\"response1\"]",null,null,null,1],
... ["wrb.fr","rpc2id","[\"some\",\"response2\"]",null,null,null,2],
... ["di",38],
... ["af.httprm",37,"5314567270682293609",6],
... ["e",4,null,null,643]]
... """
>>>
>>> # Decode response
>>> # List of tuples (index, rpcid, response)
>>> decode(raw)
[(1, 'rpc1id', ['some', 'response1']), (2, 'rpc2id', ['some', 'response2'])]
```

### Documentation

See [docs/](docs/) for more:

* [API reference](docs/api.md)
* [Response formats](docs/response-formats.md)

## Disclaimer

This project is *not* affiliated with Google. It is intended for use in accordance with [Google's Terms of Service](https://policies.google.com/terms).

## Licence

[The MIT License (MIT)](LICENSE) Copyright © 2023 Pierre Nicolas Durette

[^1]: [Deciphering Google’s mysterious 'batchexecute' system](https://kovatch.medium.com/deciphering-google-batchexecute-74991e4e446c)

[^2]: [November 2020 Google Translate API Changes](https://github.com/Boudewijn26/gTTS-token/blob/master/docs/november-2020-translate-changes.md)

CHANGELOG.md
# Changelog

## 1.0.0 (2023-04-10)


### Features

* Initial add ([#1](https://github.com/pndurette/pybatchexecute/issues/1)) ([45232d8](https://github.com/pndurette/pybatchexecute/commit/45232d8465ae87b664cf9050db761693e134e035))


### Miscellaneous Chores

* release 1.0.0 ([b645fa8](https://github.com/pndurette/pybatchexecute/commit/b645fa86574f5e96a4e43b6334218789758eccca))


gbatchexecute/test_core.py
from gbatchexecute import gBatchExecute, gBatchPayload, gBatchExecuteDecodeException
import pytest


def test_init_payload():
    """Payload parameter test: payload gets converted to list"""

    payload = gBatchPayload("abc", [123])
    url = "http://test.com"

    gbe1 = gBatchExecute(payload, url=url)
    gbe2 = gBatchExecute([payload], url=url)

    assert gbe1.payload == gbe2.payload


def test_init_url_mandatory():
    """URL parameter tests: mandatory parameters"""

    payload = gBatchPayload("abc", [123])

    with pytest.raises(AssertionError) as exc1:
        gbe1 = gBatchExecute(payload)

    with pytest.raises(AssertionError) as exc2:
        gbe2 = gBatchExecute(payload, host="abc")

    assert "'host' is required" in str(exc1.value)
    assert "'app' is required" in str(exc2.value)


def test_init_url_output():
    """URL parameter tests: building URLs from parameters"""

    payload = gBatchPayload("abc", [123])

    gbe1 = gBatchExecute(payload, host="abc", app="def")
    gbe2 = gBatchExecute(payload, host="abc", app="def", user="ghi")
    gbe3 = gBatchExecute(payload, url="jkl")

    assert gbe1.url == "https://abc/_/def/data/batchexecute"
    assert gbe2.url == "https://abc/u/ghi/_/def/data/batchexecute"
    assert gbe3.url == "jkl"


def test_init_params():
    """Query 'params' parameter tests: query 'params' and constraints"""

    payload1 = gBatchPayload("abc", [123])
    payload2 = gBatchPayload("def", [123])

    gbe1 = gBatchExecute([payload1, payload2], url="xyz")

    # 'rpcids' is a join and the order is not garanteed
    assert gbe1.params["rpcids"] == "abc,def" or gbe1.params["rpcids"] == "def,abc"

    # 'rt' is always 'c'
    assert gbe1.params["rt"] == "c"

    # '_reqid' is calculated with 'reqid' and 'idx':
    # _reqid = reqid + (idx * 100000)

    # a) explicit 'reqid' and 'idx'
    #    i.e. _reqid = 3 + (3 * 100000)
    gbe2 = gBatchExecute(payload1, url="xyz", idx=3, reqid=3)
    assert gbe2.params["_reqid"] == 3 + (3 * 100000)

    # b) 'idx' is 1 by default
    #    i.e _reqid = 2 + (1 * 100000)
    gbe2 = gBatchExecute(payload1, url="xyz", reqid=2)
    assert gbe2.params["_reqid"] == 2 + (1 * 100000)

    # c) 'reqid' is rand(1, 99999) by default
    #    i.e. _reqid = rand(1, 99999) + (2 * 100000)
    gbe3 = gBatchExecute(payload1, url="xyz", idx=2)
    assert 200001 <= gbe3.params["_reqid"] <= 299999

    # Custom params merged in
    gbe4 = gBatchExecute(payload1, url="xyz", params={"custom": 123})
    assert gbe4.params["custom"] == 123

    # Custom params setter: required keys
    gbe4 = gBatchExecute(payload1, url="xyz")

    with pytest.raises(AssertionError) as exc1:
        # missing 'rpcids'
        gbe4.params = {"rt": 2, "_reqid": 3}

    with pytest.raises(AssertionError) as exc2:
        # missing 'rt'
        gbe4.params = {"rpcids": 1, "_reqid": 3}

    with pytest.raises(AssertionError) as exc3:
        # # missing '_reqid'
        gbe4.params = {"rpcids": 1, "rt": 2}

    assert "params is missing key 'rpcids'" in str(exc1.value)
    assert "params is missing key 'rt'" in str(exc2.value)
    assert "params is missing key '_reqid'" in str(exc3.value)


def test_envelope():
    """_envelope() (or frame) tests for a given payload used to determine 'f.req'"""

    payload = gBatchPayload("abc", [123])
    gbe = gBatchExecute(payload, url="xyz")

    # 1 payload (last elem: 'generic')
    assert gbe._envelope(payload, 0) == ["abc", "[123]", None, "generic"]

    # > 1 payload (last elem: payload index)
    assert gbe._envelope(payload, 1) == ["abc", "[123]", None, "1"]


def test_freq():
    """_freq data value tests, uses _envelope()"""

    payload1 = gBatchPayload("abc", [123])
    payload2 = gBatchPayload("def", [123])

    gbe1 = gBatchExecute(payload1, url="xyz")
    gbe2 = gBatchExecute([payload1, payload2], url="xyz")

    assert gbe1._freq() == """[[["abc","[123]",null,"generic"]]]"""
    assert gbe2._freq() == """[[["abc","[123]",null,"1"],["def","[123]",null,"2"]]]"""


def test_init_data():
    """'data' tests: POST data passed to request, uses _freq"""

    payload = gBatchPayload("abc", [123])

    gbe1 = gBatchExecute(payload, url="xyz")
    assert gbe1.data == {"f.req": """[[["abc","[123]",null,"generic"]]]"""}

    # Custom data setter: required keys
    gbe4 = gBatchExecute(payload, url="xyz")

    with pytest.raises(AssertionError) as exc1:
        # missing 'f.req'
        gbe4.data = {}

    assert f"data is missing key 'f.req'" in str(exc1.value)


def test_init_headers():
    """'headers' tests: default and custom headers"""

    payload = gBatchPayload("abc", [123])

    # Default headers
    gbe1 = gBatchExecute(payload, url="xyz")
    assert gbe1.headers == {
        "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
    }

    # Additional headers
    gbe2 = gBatchExecute(payload, url="xyz", headers={"custom": 123})
    assert gbe2.headers == {
        "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
        "custom": 123,
    }

    # Custom headers setter
    gbe3 = gBatchExecute(payload, url="xyz")
    gbe3.headers = {"custom": 123}
    assert gbe3.headers == {"custom": 123}


def test_decode():
    """Test decoding of raw response string: simple responses"""

    # Simple response:
    # 1x rpcid:
    #   'abc', response: ['xyz']
    raw1 = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""

    assert gBatchExecute.decode(raw=raw1) == [(1, "abc", ["xyz"])]

    # Simple response:
    # 2x rpcid:
    #   'abc', response: ['xyz']
    #   'def', response: ['uvw']
    raw2 = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"1"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
596
[["wrb.fr","def","[\"uvw\"]\n",null,null,null,"2"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""

    assert gBatchExecute.decode(raw=raw2) == [
        (1, "abc", ["xyz"]),
        (2, "def", ["uvw"]),
    ]


def test_decode_broken_payload():
    """Test decoding of raw response string: malformed JSON response"""

    # Simple response:
    # 1x rpcid:
    #   'abc', response: malformed JSON string
    raw1 = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n bad_json_here ",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""

    with pytest.raises(gBatchExecuteDecodeException) as exc1:
        gBatchExecute.decode(raw=raw1)

    assert "Frame 1 (abc): data is not a valid JSON string" in str(exc1.value)


def test_decode_bad_payload():
    """Test decoding of raw response string: bad encoded string"""

    # Simple response:
    # Bad raw encoded string
    raw1 = r"""not a good raw"""

    with pytest.raises(gBatchExecuteDecodeException) as exc1:
        gBatchExecute.decode(raw=raw1)

    assert "Could not decode any frames" in str(exc1.value)


def test_decode_strict_empty_frame():
    """Test decoding of raw response string: strict mode with an empty frame"""

    # Simple response:
    # 1x rpcid:
    #   'abc', empty response ([])
    raw1 = r"""
)]}'

596
[["wrb.fr","abc","[]",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""
    with pytest.raises(gBatchExecuteDecodeException) as exc1:
        gBatchExecute.decode(raw=raw1, strict=True)

    assert "Frame 1 (abc): data is empty" in str(exc1.value)


def test_decode_strict_bad_inout_len():
    """Test decoding of raw response string: strict mode
    with bad expected number of frames
    """

    # Simple response:
    # 1x rpcid:
    #   'abc', response: ['xyz']
    raw1 = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"1"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""

    expected_rpcids = ["abc", "def"]
    with pytest.raises(gBatchExecuteDecodeException) as exc1:
        gBatchExecute.decode(raw=raw1, strict=True, expected_rpcids=expected_rpcids)

    assert "expected: 2, got: 1" in str(exc1.value)


def test_decode_strict_bad_inout_match():
    """Test decoding of raw response string: simple responses"""

    # Simple response:
    # 3x rpcid:
    #   'abc', response: ['xyz']
    #   'def', response: ['uvw']
    #   'ghi', response: ['rst']
    raw1 = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"1"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
596
[["wrb.fr","def","[\"uvw\"]\n",null,null,null,"2"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
596
[["wrb.fr","ghi","[\"rst\"]\n",null,null,null,"2"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""

    expected_rpcids = ["def", "def", "mno"]
    with pytest.raises(gBatchExecuteDecodeException) as exc1:
        gBatchExecute.decode(raw=raw1, strict=True, expected_rpcids=expected_rpcids)

    assert "expected: ['def', 'mno'], got: ['abc', 'def', 'ghi']" in str(exc1.value)

gbatchexecute/test_decode.py
import unittest

from gbatchexecute.decode import (BatchExecuteDecodeException,
                                  _decode_rt_compressed, _decode_rt_default,
                                  decode)


class TestDecodeRtCompressed(unittest.TestCase):
    def test_single_rpc(self):
        raw = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""
        expected_output1 = [(1, "abc", ["xyz"])]

        self.assertEqual(_decode_rt_compressed(raw), expected_output1)

    def test_multiple_rpcs(self):
        raw1 = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"1"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
596
[["wrb.fr","def","[\"uvw\"]\n",null,null,null,"2"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""

        expected_output1 = [(1, "abc", ["xyz"]), (2, "def", ["uvw"])]

        self.assertEqual(_decode_rt_compressed(raw1), expected_output1)

    def test_strict_invalid_json_response(self):
        raw = r"""
)]}'
596
[["wrb.fr","abc","[xyz]\n",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""
        with self.assertRaises(BatchExecuteDecodeException):
            _decode_rt_compressed(raw, strict=True)

    def test_strict_empty_response(self):
        raw = r"""
)]}'
596
[["wrb.fr","abc","[]\n",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""
        with self.assertRaises(BatchExecuteDecodeException):
            _decode_rt_compressed(raw, strict=True)


class TestDecodeRtDefault(unittest.TestCase):
    def test_single_rpc(self):
        raw = r"""
)]}'

[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"generic"],
["di",38],
["af.httprm",37,"5314567270682293609",6],
["e",4,null,null,643]]
"""
        expected_output = [(1, "abc", ["xyz"])]

        self.assertEqual(_decode_rt_default(raw), expected_output)

    def test_multiple_rpcs(self):
        raw1 = r"""
)]}'

[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,1],
["wrb.fr","def","[\"uvw\"]\n",null,null,null,2],
["di",38],
["af.httprm",37,"5314567270682293609",6],
["e",4,null,null,643]]
"""

        expected_output1 = [(1, "abc", ["xyz"]), (2, "def", ["uvw"])]

        self.assertEqual(_decode_rt_default(raw1), expected_output1)

    def test_strict_invalid_json_response(self):
        raw = r"""
)]}'

[["wrb.fr","abc","[xyz]\n",null,null,null,"generic"],
["di",38],
["af.httprm",37,"5314567270682293609",6],
["e",4,null,null,643]]
"""
        with self.assertRaises(BatchExecuteDecodeException):
            _decode_rt_default(raw, strict=True)

    def test_strict_empty_response(self):
        raw = r"""
)]}'

[["wrb.fr","abc","[]\n",null,null,null,"generic"],
["di",38],
["af.httprm",37,"5314567270682293609",6],
["e",4,null,null,643]]
"""
        with self.assertRaises(BatchExecuteDecodeException):
            _decode_rt_default(raw, strict=True)


class TestDecode(unittest.TestCase):
    def test_valid_rt_compressed(self):
        raw = r"""
)]}'

596
[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6]
]
26
[["e",4,null,null,643]
]
"""
        expected_output = [(1, "abc", ["xyz"])]
        self.assertEqual(decode(raw, rt="c"), expected_output)

    def test_valid_rt_default(self):
        raw = r"""
)]}'

[["wrb.fr","abc","[\"xyz\"]\n",null,null,null,"generic"]
,["di",38]
,["af.httprm",37,"5314567270682293609",6],
["e",4,null,null,643]]
"""
        expected_output = [(1, "abc", ["xyz"])]
        self.assertEqual(decode(raw, rt=None), expected_output)

    def test_invalid_rt(self):
        with self.assertRaises(ValueError):
            decode("test", rt="invalid")

    def test_strict_true_expected_rpcids(self):
        raw = r"""
)]}'

123
[["wrb.fr","abc","[\"xyz\"]",null,null,null,"1"]]
123
[["wrb.fr","def","[\"uvw\"]",null,null,null,"2"]]
"""
        expected_output = [(1, "abc", ["xyz"]), (2, "def", ["uvw"])]
        self.assertEqual(
            decode(raw, rt="c", strict=True, expected_rpcids=["abc", "def"]),
            expected_output,
        )

    def test_strict_true_mismatched_rpcids(self):
        raw = r"""
)]}'

123
[["wrb.fr","abc","[\"xyz\"]",null,null,null,"1"]]
123
[["wrb.fr","ijk","[\"uvw\"]",null,null,null,"2"]]
"""
        with self.assertRaises(BatchExecuteDecodeException):
            decode(raw, rt="c", strict=True, expected_rpcids=["abc", "def"])


if __name__ == "__main__":
    unittest.main()


gbatchexecute/test_encode.py
import unittest
from typing import List

from gbatchexecute.encode import BatchExecuteFunction, PreparedBatchExecute


class TestPreparedBatchExecuteFunctions(unittest.TestCase):
    def setUp(self):
        self.url_params = {"host": "uvw", "app": "xyz"}

    def test_set_valid_functions(self):
        valid_functions: List[BatchExecuteFunction] = [
            {"rpcid": "function1", "args": [1, 2]},
            {"rpcid": "function2", "args": ["a", "b"]},
        ]
        pbe = PreparedBatchExecute(valid_functions, **self.url_params)
        self.assertEqual(pbe.functions, valid_functions)

    def test_set_invalid_functions_not_list(self):
        with self.assertRaises(ValueError):
            PreparedBatchExecute("not a list", **self.url_params)

    def test_validate_function_not_dict(self):
        pbe = PreparedBatchExecute([], **self.url_params)
        with self.assertRaises(ValueError):
            pbe._validate_function("not a dictionary")

    def test_validate_function_missing_rpcid(self):
        pbe = PreparedBatchExecute([], **self.url_params)
        invalid_function: BatchExecuteFunction = {"args": [1, 2]}
        with self.assertRaises(ValueError):
            pbe._validate_function(invalid_function)

    def test_validate_function_missing_args(self):
        pbe = PreparedBatchExecute([], **self.url_params)
        invalid_function: BatchExecuteFunction = {"rpcid": "function1"}
        with self.assertRaises(ValueError):
            pbe._validate_function(invalid_function)

    def test_validate_function_invalid_rpcid_type(self):
        pbe = PreparedBatchExecute([], **self.url_params)
        # rpcid must be a string
        invalid_function: BatchExecuteFunction = {"rpcid": 123, "args": [1, 2]}
        with self.assertRaises(ValueError):
            pbe._validate_function(invalid_function)

    def test_validate_function_invalid_args_type(self):
        pbe = PreparedBatchExecute([], **self.url_params)
        # args must be a list
        invalid_function: BatchExecuteFunction = {
            "rpcid": "function1",
            "args": "not a list",
        }
        with self.assertRaises(ValueError):
            pbe._validate_function(invalid_function)


class TestPreparedBatchExecuteURL(unittest.TestCase):
    def setUp(self):
        self.functions = [{"rpcid": "abc", "args": [123]}]

    def test_set_valid_url(self):
        pbe = PreparedBatchExecute(self.functions, host="abc", app="def")
        self.assertEqual(pbe.url, "https://abc/_/def/data/batchexecute")

    def test_set_valid_url_with_user(self):
        pbe = PreparedBatchExecute(self.functions, host="abc", app="def", user="ghi")
        self.assertEqual(pbe.url, "https://abc/u/ghi/_/def/data/batchexecute")


class TestPreparedBatchExecuteParams(unittest.TestCase):
    def setUp(self):
        self.function1 = {"rpcid": "abc", "args": [123]}
        self.function2 = {"rpcid": "def", "args": [123]}

        self.url_params = {"host": "uvw", "app": "xyz"}

    def test_rpcids(self):
        """'rpcids' parameter tests: join of 'rpcid' values"""
        pbe = PreparedBatchExecute([self.function1, self.function2], **self.url_params)
        # 'rpcids' is a join and the order is not garanteed
        self.assertIn(pbe.params["rpcids"], ["abc,def", "def,abc"])

    def test_reqid(self):
        """'_reqid' parameter tests: 'reqid' + (index * 100000)"""
        # a) explicit 'reqid' and 'idx'
        #    i.e. _reqid = 1000 + (3 * 100000)
        pbe1 = PreparedBatchExecute(
            [self.function1], reqid=1000, index=3, **self.url_params
        )
        self.assertEqual(pbe1.params["_reqid"], 1000 + (3 * 100000))

        # b) 'index' is 0 by default
        #    i.e _reqid = 1000 + (0 * 100000)
        pbe2 = PreparedBatchExecute([self.function1], reqid=1000, **self.url_params)
        self.assertEqual(pbe2.params["_reqid"], 1000 + (0 * 100000))

        # c) 'reqid' is rand(1000, 9999) by default
        #    i.e. _reqid = rand(1000, 9999) + (2 * 100000)
        gbe3 = PreparedBatchExecute([self.function1], index=2, **self.url_params)
        self.assertTrue(200000 <= gbe3.params["_reqid"] <= 209999)

    def test_rt(self):
        """'rt' is set (if not None)"""
        pbe1 = PreparedBatchExecute([self.function1], rt="a", **self.url_params)
        pbe2 = PreparedBatchExecute([self.function1], rt=None, **self.url_params)
        assert pbe1.params["rt"] == "a"
        assert "rt" not in pbe2.params


class TestPreparedBatchExecuteExecuteData(unittest.TestCase):
    def setUp(self):
        self.function1 = {"rpcid": "abc", "args": [123]}
        self.function2 = {"rpcid": "def", "args": [123]}

        self.url_params = {"host": "uvw", "app": "xyz"}

    def test_data_freq(self):
        """'data' data value tests"""
        pbe1 = PreparedBatchExecute([self.function1], **self.url_params)
        pbe2 = PreparedBatchExecute([self.function1, self.function2], **self.url_params)

        # 1 function (last elem: 'generic')
        assert pbe1.data["f.req"] == '[[["abc","[123]",null,"generic"]]]'

        # > 1 function (last elem: function index)
        assert (
            pbe2.data["f.req"]
            == '[[["abc","[123]",null,"1"],["def","[123]",null,"2"]]]'
        )


class TestPreparedBatchExecuteExecuteHeaders(unittest.TestCase):
    def setUp(self):
        self.function = {"rpcid": "abc", "args": [123]}
        self.url_params = {"host": "uvw", "app": "xyz"}

    def test_headers(self):
        """'headers' data value tests"""
        pbe = PreparedBatchExecute([self.function], **self.url_params)
        assert pbe.headers == {
            "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
        }


if __name__ == "__main__":
    unittest.main()


pybatchexecute/encode.py
import json
import random
from typing import List, TypedDict

__all__ = ["PreparedBatchExecute"]


class BatchExecuteRpc(TypedDict):
    """Type hint for a batchexecute RPC"""

    rpcid: str
    args: List[str]


class PreparedBatchExecute(object):
    """A prepared ``batchexecute`` RPC

    This class prepares a ``batchexecute`` RPC for sending.
    It provides the URL, URL parameters, headers and POST data for the request,
    which can be used to send a POST request using any HTTP library.

    **Properties**:
      * ``url`` _str_ - The URL to send the request to
      * ``params`` _dict_ - The URL parameters to send with the request
      * ``data`` _dict_ - The POST data to send with the request
      * ``headers`` _dict_ -  The request headers to send with the request

    """

    def __init__(
        self,
        rpcs: List[BatchExecuteRpc],
        host: str,
        app: str,
        user: str = None,
        reqid: int = None,
        index: int = 0,
        rt: str = None,
    ) -> None:
        """Prepare a ``batchexecute`` request

        _The URL_: The URL is built from the ``host``, ``app``, and ``user`` parameters
        (each parameter can be obtained by inspecting an existing request):
        ``https://{host}/_/{app}/data/batchexecute`` or
        ``https://{host}/u/{user}/_/{app}/data/batchexecute`` (if ``user`` is not ``None``)

        _Notes about ``reqid`` and ``index``_:
        The ``_reqid`` parameter is calculated from the ``reqid`` and ``index`` parameters.
        It is meant for when many ``PreparedBatchExecute`` are prepared to be sent sequentially:
        each prepared request should have an incremental ``index`` and use the same ``reqid``.
        It is calculated as ``reqid + (index * 100000)`` (four digits, with ``100000`` added
        onto it with each subsequent request, e.g. ``1234`` , ``101234`` , ``201234``, etc.)

        _Notes about ``rt``_:
        The ``rt`` parameter defines the response. It can be ``"c"`` (compressed),
        ``"b"`` (ProtoBuf), or ``None`` (JSON-ish). See [``decode`` module methods](#decode)
        for more information.

        Args:
            rpcs (list): A list of rpcs to execute. Each RPC is a dictionary with 2 keys:
                * ``rpcid``: The ``rpcid`` of the RPC to execute
                * ``args``: A list of arguments to pass to the RPC
            host (str): The host to send the request to
            app (str): The app to send the request to
            user (str): The user to send the request to (default: ``None``)
            reqid (int): The request ID. Must be a four digit number (default: random if ``None``)
            index (int): The index of this ``PreparedBatchExecute``, in case many are prepared to
                be sent sequentially. Each prepared request should have an incremental ``index``
                and use the same ``reqid`` (default: ``0``)
            rt (str): The response type. Can be``"c"`` (compressed),
                ``"b"`` (ProtoBuf), or ``None`` (JSON) (default: ``None``)

        Raises:
            ValueError: If ``reqid`` is not a four digit number
            ValueError: If any RPC is of an invalid format

        """
        # rpcs
        self.rpcs = rpcs

        # URL components
        self.host = host
        self.app = app
        self.user = user

        # Paramemter components
        self.reqid = reqid
        self.index = index
        self.rt = rt

    @property
    def headers(self) -> dict:
        """Get request headers"""
        return {
            "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
        }

    @property
    def rpcs(self) -> List[BatchExecuteRpc]:
        """Get the rpcs to be called"""
        return self._rpcs

    @rpcs.setter
    def rpcs(self, rpcs: List[BatchExecuteRpc]) -> None:
        """Set the rpcs to be called"""
        # rpcs must be a list
        if not isinstance(rpcs, list):
            raise ValueError("'rpcs' must be a list")

        # Validate each RPC
        for fct in rpcs:
            self._validate_rpc(fct)

        self._rpcs = rpcs

    @property
    def url(self) -> str:
        """Get the URL"""
        if not self.user:
            return f"https://{self.host}/_/{self.app}/data/batchexecute"
        else:
            return f"https://{self.host}/u/{self.user}/_/{self.app}/data/batchexecute"

    @property
    def reqid(self) -> int:
        """Get the request ID"""
        return self._reqid

    @reqid.setter
    def reqid(self, reqid) -> None:
        """Set the request ID"""
        if reqid and not 1000 <= reqid <= 9999:
            raise ValueError("'reqid' must be a four digit positive integer")
        elif reqid:
            self._reqid = reqid
        else:
            self._reqid = random.randrange(1000, 9999)

    @property
    def params(self) -> dict:
        """Build the URL parameters for a ``batchexecute`` RPC"""

        params = {
            "rpcids": ",".join(set([fct["rpcid"] for fct in self.rpcs])),
            "_reqid": self._reqid + (self.index * 100000),
        }

        if self.rt:
            params["rt"] = self.rt

        return params

    @property
    def data(self) -> dict:
        """Build the POST data for a ``batchexecute`` RPC

        The data is a dictionary with a single key ``f.req``,
        which looks like, as stringified JSON:

        ```
        [ # Outer array
            [ # List of rpcs
                [ # RPC 'envelope'
                    "<RPC rpcid>",
                    "<RPC args stringified JSON>",
                    null,
                    "<index of RPC or 'generic'>"
                ],
                [...]
            ]
        ]
        ```

        Returns:
            dict: The POST data

        """

        def _envelope(rpc: BatchExecuteRpc, rpc_idx: int = 0) -> list:
            """Build an 'envelope' of a RPC

            Args:
                rpc (dict): A dictionary with 2 keys:
                    * ``rpcid``: The ``rpcid`` of the RPC to execute
                    * ``args``: A list of arguments to pass to the RPC
                rpc_idx (int): The index of the RPC in the list of rpcs

            Returns:
                list: The RPC 'envelope', of the form::

                    [
                        "<RPC rpcid>",
                        "<RPC args stringified JSON>",
                        None,
                        <index of RPC or "generic" if index is 0>
                    ]

            """

            return [
                rpc["rpcid"],
                json.dumps(rpc["args"], separators=(",", ":")),
                None,
                str(rpc_idx) if rpc_idx > 0 else "generic",
            ]

        # List of RPC 'envelopes' to request
        freq = []

        for fct_idx, fct in enumerate(self.rpcs, start=1):
            if len(self.rpcs) == 1:
                fct_idx = 0

            freq.append(_envelope(fct, fct_idx))

        # Wrap in outer array, dump to JSON string
        freq = json.dumps([freq], separators=(",", ":"))

        return {"f.req": freq}

    def _validate_rpc(self, rpc: BatchExecuteRpc) -> None:
        """Validate a RPC format for a ``batchexecute`` RPC

        Args:
            rpc (dict): A dictionary with 2 keys:
                rpcid: The ``rpcid`` of the RPC to execute
                args: A list of arguments to send to the ``rpcid`` RPC

        """
        if not isinstance(rpc, dict):
            raise ValueError("RPC must be a dictionary")

        if not "rpcid" in rpc:
            raise ValueError("RPC must have a 'rpcid' key")

        if not "args" in rpc:
            raise ValueError("RPC must have an 'args' key")

        if not isinstance(rpc["rpcid"], str):
            raise ValueError("RPC 'rpcid' must be a string")

        if not isinstance(rpc["args"], list):
            raise ValueError("RPC 'args' must be a list")
